src/Tools/Replay/Replay.cs

```diff
@@ -0,0 +1,271 @@
+            index++;
+        }
+
+        var completedTask = await Task.WhenAny(tasks).ConfigureAwait(false);
+        if (!tasks.Remove(completedTask))
+        {
+            throw new Exception("Task was not in the list");
+        }
+
+        var buildData = await completedTask.ConfigureAwait(false);
+        yield return buildData;
+    } while (tasks.Count > 0);
+
+    string GetOutputName(CompilerCall compilerCall)
+    {
+        string name;
+        if (compilerCall.Kind is CompilerCallKind.Regular)
+        {
+            name = $"{compilerCall.ProjectFileName}-{compilerCall.TargetFramework}";
+        }
+        else
+        {
+            name = $"{compilerCall.ProjectFileName}-{compilerCall.TargetFramework}-{compilerCall.Kind}";
+        }
+
+        if (!outputSet.Add(name))
+        {
+            name = $"{name}-{outputSet.Count}";
+        }
+
+        return name;
+    }
+}
+
+static async Task<BuildData> BuildAsync(
+    ReplayOptions options,
+    CompilerCall compilerCall,
+    string outputName,
+    CompilerServerLogger compilerServerLogger,
+    CancellationToken cancellationToken)
+{
+    var args = compilerCall.GetArguments();
+    var outputDirectory = Path.Combine(options.OutputDirectory, outputName);
+    Directory.CreateDirectory(outputDirectory);
+    RewriteOutputPaths(outputDirectory, args);
+
+    var request = BuildServerConnection.CreateBuildRequest(
+        outputName,
+        compilerCall.IsCSharp ? RequestLanguage.CSharpCompile : RequestLanguage.VisualBasicCompile,
+        args.ToList(),
+        workingDirectory: compilerCall.ProjectDirectory,
+        tempDirectory: options.TempDirectory,
+        keepAlive: null,
+        libDirectory: null);
+    var response = await BuildServerConnection.RunServerBuildRequestAsync(
+        request,
+        options.PipeName,
+        options.ClientDirectory,
+        compilerServerLogger,
+        cancellationToken).ConfigureAwait(false);
+    return new BuildData(compilerCall, response);
+}
+
+static void RewriteOutputPaths(string outputDirectory, string[] args)
+{
+    var comparison = StringComparison.OrdinalIgnoreCase;
+    var hashSet = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+    for (var i = 0; i < args.Length; i++)
+    {
+        var line = args[i];
+        var isOption = line.Length > 0 && line[0] is '-' or '/';
+        if (!isOption)
+        {
+            continue;
+        }
+
+        line = line.Substring(1);
+        // Map all of the output items to the build output directory
+        if (line.StartsWith("out", comparison) ||
+            line.StartsWith("refout", comparison) ||
+            line.StartsWith("doc", comparison) ||
+            line.StartsWith("errorlog", comparison))
+        {
+            var index = line.IndexOf(':');
+            var argValue = line.AsSpan().Slice(index + 1).ToString();
+
+            string fileName;
+            if (Path.IsPathRooted(argValue))
```

Comment:
Is this going to have a cross-platform problem? IE, when I try to use this on Linux to replay a binary log from Windows? Or do we explicitly only support replaying on the same machine?

Reply:
This is only replaying on the same machine. Will clarify in the docs.